driver/compile_common.ml

```diff
@@ -113,15 +113,16 @@ let typecheck_impl i parsetree =
 
 let implementation info ~backend =
   Profile.record_call info.source_file @@ fun () ->
-  let sufs = if info.native then [ cmx; obj ] else [ cmo ] in
+  let exceptionally () =
+    let sufs = if info.native then [ cmx; obj ] else [ cmo ] in
+    List.iter (fun suf -> remove_file (suf info)) sufs;
+  in
+  Misc.try_finally ?always:None ~exceptionally @@ fun () ->
   let parsed = parse_impl info in
   if Clflags.(should_stop_after Compiler_pass.Parsing) then () else begin
     let typed = typecheck_impl info parsed in
     if Clflags.(should_stop_after Compiler_pass.Typing) then () else begin
-      let exceptionally () =
-        List.iter (fun suf -> remove_file (suf info)) sufs;
-      in
-      Misc.try_finally ~exceptionally (fun () -> backend info typed)
+      backend info typed
     end;
   end;
   Warnings.check_fatal ();
```

Comment:
I would wrap the term after `@@` with `begin .. end`, otherwise it is not immediately clear if it includes `Warnings.check_fatal ()` or not, which is just the point.

Reply:
I'm not fond of the suggestion, because (1) to me it's clear that this idiom goes until the end of the expression (this comes from the precedence of `fun`), and (2) it makes the style inconsistent with the call to `Profile.record_call` above in the function. But on the other hand, (1) some people don't like `@@ fun` and (2) being extra careful about scoping of effects is reasonable. So I'll change to `(fun ()-> ...)`.

Reply:
Fair enough!